File: 7/list-7.go
package main

import (
	"fmt"
)

type Escola struct {
	local       string
	professores []*Professor
}

func (e *Escola) adicionarProfessor(p *Professor) {
	e.professores = append(e.professores, p)
}

type Professor struct {
	nome    string
	escolas []*Escola
}

func (p *Professor) adicionarEscola(e *Escola) {
	p.escolas = append(p.escolas, e)
}

func main() {
	professor1 := &Professor{nome: "Carlos"}
	professor2 := &Professor{nome: "Marcio"}

	escola1 := &Escola{local: "Escola A"}
	escola2 := &Escola{local: "Escola B"}

	escola1.adicionarProfessor(professor1)
	escola1.adicionarProfessor(professor2)
	escola2.adicionarProfessor(professor1)
	escola2.adicionarProfessor(professor2)

	professor1.adicionarEscola(escola1)
	professor1.adicionarEscola(escola2)

	fmt.Printf("Escola: %s tem professores: ", escola1.local)
	for _, prof := range escola1.professores {
		fmt.Printf("%s ", prof.nome)
	}
	fmt.Println()

	fmt.Printf("Escola: %s tem professores: ", escola2.local)
	for _, prof := range escola2.professores {
		fmt.Printf("%s ", prof.nome)
	}
	fmt.Println()

	fmt.Printf("Professor: %s leciona nas escolas: ", professor1.nome)
	for _, escola := range professor1.escolas {
		fmt.Printf("%s ", escola.local)
	}
	fmt.Println()
}

File: 7/list-7.py
class Escolas:
    def __init__(self,local):
        self.local = local
        self.professores = []

    def adicionar_professor(self, professor):
        self.professores.append(professor)

class Professores:
    def __init__(self, nome):
        self.nome = nome
        self.escola = []

    def adicionar_escola(self, escola):
        self.escola.append(escola)

def main():
    professor1 = Professores("Carlos")
    professor2 = Professores("Marcio")

    escola1 = Escolas("Escola A")
    escola2 = Escolas("Escola B")

    escola1.adicionar_professor(professor1)
    escola1.adicionar_professor(professor2)
    escola2.adicionar_professor(professor1)
    escola2.adicionar_professor(professor2)

    professor1.adicionar_escola(escola1)
    professor1.adicionar_escola(escola2)

    print(f'Escola: {escola1.local} tem professores {[prof.nome for prof in escola1.professores]}')
    print(f'Escola: {escola2.local} tem professores {[prof.nome for prof in escola2.professores]}')

    print(f'Profersor:{professor1.nome} leciona nas escolas {[escola.local for escola in professor1.escola]}')

if __name__ == "__main__":
    main()
